01_Data/src/layers/sequelize/repository/Certificate.ts

```diff
@@ -7,28 +7,30 @@ import { ICertificateRepository } from '../../../interfaces';
 import { Certificate } from '../model/Certificate';
 import { SystemConfig } from '@citrineos/base';
 import { Sequelize } from 'sequelize-typescript';
-import { Logger, ILogObj } from 'tslog';
+import { ILogObj, Logger } from 'tslog';
 
 export class SequelizeCertificateRepository extends SequelizeRepository<Certificate> implements ICertificateRepository {
   constructor(config: SystemConfig, logger?: Logger<ILogObj>, sequelizeInstance?: Sequelize) {
     super(config, Certificate.MODEL_NAME, logger, sequelizeInstance);
   }
 
   async createOrUpdateCertificate(certificate: Certificate): Promise<Certificate> {
-    const [storedCert, _certCreated] = await Certificate.upsert({
-      signedBy: certificate.signedBy,
-      serialNumber: certificate.serialNumber,
-      keyLength: certificate.keyLength,
-      organizationName: certificate.organizationName,
-      commonName: certificate.commonName,
-      validBefore: certificate.validBefore,
-      signatureAlgorithm: certificate.signatureAlgorithm,
-      countryName: certificate.countryName,
-      isCA: certificate.isCA,
-      pathLen: certificate.pathLen,
-      certificateFileId: certificate.certificateFileId,
-      privateKeyFileId: certificate.privateKeyFileId,
+    return await this.s.transaction(async (transaction) => {
+      // TODO what to search by
+      const savedCert = await this.s.models[Certificate.MODEL_NAME].findOne({
+        where: {
+          serialNumber: certificate.serialNumber,
+          organizationName: certificate.organizationName,
+          commonName: certificate.commonName,
+        },
+        transaction,
+      });
+      if (!savedCert) {
+        await certificate.save({ transaction });
+      } else {
+        await certificate.update(certificate, { transaction });
+      }
+      return await certificate.reload({ transaction });
```

Comment:
Do we need to send the emit event here? Sry I am a bit confused about when to emit and when not.

Reply:
Hmm. I think you are correct that it should emit here and that we should rather use CRUD repository. Before, it use the models like I did here. Which didn't emit. But I think we should be consistent and use the CRUD repo methods. Making the adjustment.

Reply:
I have gone over my changes again. So the pattern should be that if we are directly creating/updating we use the CRUD repo to emit. If we instead are creating, then attaching more info to the same object, we would want to emit the full object, not just the half one where we don't have all fields set yet. Example would be the Boot one.